OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,421 @@
+								Game.CosmeticRandom.Next(1, 4),
+								Game.CosmeticRandom.Next(0, 3) - 1,
+								Game.CosmeticRandom.Next(10, 20) / 10f,
+								Game.CosmeticRandom.Next(10, 15) / 10f,
+								Game.CosmeticRandom.Next(1, 25) / 1000f,
+								TempSwingDirection,
+								Game.CosmeticRandom.Next(10, 15) / 10f,
+								Color.FromArgb(255 - Game.CosmeticRandom.Next(0, 51), tempLittleBlue, tempLittleBlue, tempBrightness),
+								Color.Pink
+							)
+						);
+					}
+					if (info.WeatherVariants == "rain")
+					{
+						var tempBrightness = 175 + Game.CosmeticRandom.Next(0, 6) * 10;
+						var tempLittleBlue = tempBrightness - Game.CosmeticRandom.Next(60, 90);
+
+						DroppingParticleList.Add(
+							new droppingParticle(
+								Game.CosmeticRandom.Next(0, Game.Renderer.Resolution.Width + 1),
+								Game.CosmeticRandom.Next(0, rangeY + 1),
+								Game.CosmeticRandom.Next(1, 2),
+								0,
+								Game.CosmeticRandom.Next(80, 120) / 10f,
+								0,
+								0,
+								false,
+								0,
+								Color.FromArgb(255 - Game.CosmeticRandom.Next(0, 21), tempLittleBlue, tempLittleBlue, tempBrightness),
+								Color.FromArgb(255, tempLittleBlue, tempLittleBlue, tempBrightness)
+							)
+						);
+					}
+					if (info.WeatherVariants == "d2kSand")
+					{
+						bool TempSwingDirection = Game.CosmeticRandom.Next(0, 2) == 0 ? true : false;
+						var tempBrightness = Game.CosmeticRandom.Next(0, 61);
+
+						float tempGravitiy = 0;
+						if (Game.CosmeticRandom.Next(0, 4) == 1)
+							tempGravitiy = Game.CosmeticRandom.Next(5, 10) / 10f;
+						else 
+							tempGravitiy = - Game.CosmeticRandom.Next(5, 10) / 10f ;
+
+						DroppingParticleList.Add(
+							new droppingParticle(
+								Game.CosmeticRandom.Next(0, Game.Renderer.Resolution.Width + 1),
+								Game.CosmeticRandom.Next(0, rangeY + 1),
+								2,
+								0,
+								tempGravitiy,
+								0,
+								0,
+								TempSwingDirection,
+								0,
+								Color.FromArgb(255 - Game.CosmeticRandom.Next(0, 21), 248 - tempBrightness, 198 - tempBrightness, 75 - tempBrightness),
+								Color.Pink
+							)
+						);
+					}
+				}
+			}
+		}
+
+		void DroppingParticlesCountLogic(WorldRenderer wr)
+		{
+			// Logic 
+			if (DrPaCountFader == DrPaCountFaderType.FadeIn && DroppingParticleList.Count >= MaxDrPaCount)
+				DrPaCountFader = DrPaCountFaderType.Hold;
+			else if (DrPaCountFader == DrPaCountFaderType.FadeOut && DroppingParticleList.Count == 0)
+				DrPaCountFader = DrPaCountFaderType.Hold;
+
+			// Do the fade functions
+			if (DrPaCountFader == DrPaCountFaderType.FadeIn)
+				FadeInCountDrPa(wr);
+			else if (DrPaCountFader == DrPaCountFaderType.FadeOut)
+				FadeOutCountDoPa(wr);
+		}
+
+		void FadeInCountDrPa(WorldRenderer wr)
+		{				 
+			AddADroppingParticle(1, 0, 100);
+
+			// Remove DroppingParticles, which is getting replaced from the top to the bottom, when scrolling down as long as FadeIn
+			// to avoid a gap in the middle of the screen
```

Comment:
Because I'd like to avoid alot adding and deleting of particles, I'm moveing them to the opposide screen edge ( like in the old Asteroids game).
I fade the count of particles to maxcount by spawning them at y= 0 and they can fall down slowly. The player can now decide to scroll up even the fade in is not complete. Means the particles would get moved from the top to the bottom. and that looks ugly, because of the middle is empty. This part avoids that.